solidity/contracts/bridge/Bridge.sol

```diff
@@ -795,10 +935,500 @@ contract Bridge is Ownable {
-    //      selectively, based on deposits they have earlier received.
-    //      (UPDATE PR #90: Is it still the case since amounts are inferred?)
+    /// @notice Requests redemption of the given amount from the specified
+    ///         wallet to the redeemer Bitcoin output script hash.
+    /// @param walletPubKeyHash The 20-byte wallet public key hash (computed
+    ///        using HASH160 opcode)
+    /// @param mainUtxo Data of the wallet's main UTXO, as currently known on
+    ///        the Ethereum chain
+    /// @param redeemerOutputHash The 20-byte (P2PKH, P2WPKH, or P2SH) or
+    ///        32-byte (P2WSH) output hash that will be used to lock
+    ///        redeemed BTC
+    /// @param amount Requested amount in satoshi. This is also the TBTC amount
+    ///        that is taken from redeemer's balance in the Bank upon request.
+    ///        Once the request is handled, the actual amount of BTC locked
+    ///        on the redeemer output hash will be always lower than this value
+    ///        since the treasury and Bitcoin transaction fees must be incurred.
+    ///        The minimal amount satisfying the request can be computed as:
+    ///        `amount - redemptionTreasuryFee - redemptionTxMaxFee`.
+    ///        Fees values are taken at the moment of request creation.
+    /// @dev Requirements:
+    ///      - Wallet behind `walletPubKeyHash` cannot be marked as faulty
+    ///      - `mainUtxo` components must point to the recent main UTXO
+    ///        of the given wallet, as currently known on the Ethereum chain.
+    ///      - `redeemerOutputHash` must be proper 20-byte or 32-byte BTC hash
+    ///      - `redeemerOutputHash` cannot be the same as `walletPubKeyHash`
+    ///      - `amount` must be above or equal the `redemptionDustThreshold`
+    ///      - Given `walletPubKeyHash` and `redeemerOutputHash` pair can be
+    ///        used for only one pending request at the same time
+    ///      - Wallet must have enough Bitcoin balance to proceed the request
+    ///      - Redeemer must make an allowance in the Bank that the Bridge
+    ///        contract can spend the given `amount`.
+    function requestRedemption(
+        bytes20 walletPubKeyHash,
+        BitcoinTx.UTXO calldata mainUtxo,
+        bytes calldata redeemerOutputHash,
+        uint64 amount
+    ) external {
+        require(!faultyWallets[walletPubKeyHash], "Wallet marked as faulty");
+
+        bytes32 mainUtxoHash = mainUtxos[walletPubKeyHash];
+        require(mainUtxoHash != bytes32(0), "No main UTXO for given wallet");
+        require(
+            keccak256(
+                abi.encodePacked(
+                    mainUtxo.txHash,
+                    mainUtxo.txOutputIndex,
+                    mainUtxo.txOutputValue
+                )
+            ) == mainUtxoHash,
+            "Invalid main UTXO data"
+        );
+
+        // TODO: Check if `walletPubKeyHash` should be validated in some other ways.
```

Comment:
Hmmm... like what? Nothing comes to mi mind, do you have something?

Reply:
Maybe some requirements regarding wallet age? As far I remember only the oldest wallets should redeem. The separate question is whether we want to force that on the contract side. This is why that TODO is here.

Reply:
Let's improve this TODO then:

```suggestion
        // TODO: Confirm if `walletPubKeyHash` should be validated by checking if it is the
        // oldest one who can handle the request. This will be suggested by the dApp
        // but may not be respected by users who interact directly with the contract.
        // Do we need to enforce it here?
        // One option is not to enforce it, to save on gas, but if we see this rule
        // is not respected, upgrade Bridge contract to require it.
```

Reply:
See https://github.com/keep-network/tbtc-v2/pull/128/commits/b96f1ca4c56abcc4404ff0e78977ed54ff8d30c6